digest/src/core_api/rt_variable.rs

```diff
@@ -145,6 +149,68 @@ where
     }
 }
 
+impl<T> SerializableState for RtVariableCoreWrapper<T>
+where
+    T: VariableOutputCore + UpdateCore + SerializableState,
+    T::BlockSize: IsLess<U256>,
+    Le<T::BlockSize, U256>: NonZero,
+    T::SerializedStateSize: Add<U9>,
+    <T::SerializedStateSize as Add<U9>>::Output: Add<T::BlockSize>,
+    <<T::SerializedStateSize as Add<U9>>::Output as Add<T::BlockSize>>::Output: ArrayLength<u8>,
+{
+    type SerializedStateSize = Sum<Sum<T::SerializedStateSize, U9>, T::BlockSize>;
+
+    fn serialize(&self) -> SerializedState<Self> {
+        let mut serialized_state = SerializedState::<Self>::default();
+        let serialization_buffer = serialized_state.as_mut_slice();
+
+        let serialized_core = self.core.serialize();
+        let (buffer, serialization_buffer) =
+            serialization_buffer.split_at_mut(serialized_core.len());
+        buffer.copy_from_slice(serialized_core.as_slice());
+
+        let pos = self.buffer.get_pos().try_into().unwrap();
+        let (buffer, serialization_buffer) = serialization_buffer.split_at_mut(size_of_val(&pos));
+        buffer[0] = pos;
+
+        let data = self.buffer.clone().pad_with_zeros().clone();
+        let (buffer, serialization_buffer) = serialization_buffer.split_at_mut(data.len());
+        buffer.copy_from_slice(data.as_slice());
+
+        let output_size = self.output_size.try_into().unwrap();
+        serialization_buffer[0] = output_size;
+
+        serialized_state
+    }
+
+    fn deserialize(
+        serialized_state: &SerializedState<Self>,
+    ) -> Result<Self, DeserializeStateError> {
+        let serialization_buffer = serialized_state.as_slice();
+
+        let (buffer, serialization_buffer) =
+            serialization_buffer.split_at(T::SerializedStateSize::USIZE);
+        let serialized_core = SerializedState::<T>::from_slice(buffer);
+
+        let (buffer, serialization_buffer) = serialization_buffer.split_at(1);
+        let pos = buffer[0].into();
+        if pos > T::BlockSize::USIZE {
+            return Err(DeserializeStateError);
+        }
+
+        let (block_buffer, serialization_buffer) =
+            serialization_buffer.split_at(T::BlockSize::USIZE);
+
+        let output_size = serialization_buffer[0].into();
+
+        Ok(Self {
+            core: T::deserialize(serialized_core)?,
+            buffer: BlockBuffer::new(&block_buffer[..pos]),
```

Comment:
Note that if the buffer is "eager" and `pos` is equal to block size, `new` will panic. I guess the simplest solution will be to add `try_new` method to `BlockBuffer`.

Reply:
https://github.com/RustCrypto/utils/pull/799